[SD-1145] Use refined NonNegative instead of hand rolled Natural

jedesah opened the pull request:
If this is viewed as an improvement. I can do the same thing for Positive and we can use some of the other features of refined like typesafe regexes and such.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,17 @@
+import eu.timepit.refined.refineT
+import scalaz.{Equal, Show}
+
+package object numeric {
```
How about an alias here, `Natural = Long @@ NonNegative`? This will eliminate many of the renames in this PR as we're not changing the meaning, just the implementation of this concept.

jedesah replied:
Ya, hesitated because the information on whether it's an `Int`, `Long` `Short` is kinda of hidden, but I think it would be helpful. I also thought about creating an alias `Natural == NonNegative`.

wemrysi replied:
Also, is it possible to create `val Natural = refineMT[NonNegative].apply[Long]` or something to that effect?

jedesah replied:
I think it should be possible. I will try it out.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,17 @@
+import scalaz.{Equal, Show}
+
+package object numeric {
+  def add[T:scala.Numeric](a:T @@ NonNegative, b: T @@ NonNegative): T @@ NonNegative =
```
Can we create a new `object natural` inside `quasar.fp.numeric` to hold these? We can also implement `Monoid[Natural]` (or `[A: Numeric]Monoid[A @@ NonNegative]`) instead of the one-off `add` method.

jedesah replied:
Sure. I tried making a `Monoid[A @@ NonNegative]` but ran into a bizarre issue where I was unable to define an `append` that "overrided" the abstract member `append` from `Monoid`. Possibly some problem with `shapeless.tag.@@` or `refined` or I was just being dumb, not sure.

I also tried something like `def add[N <: Nat,M M: Nat,T](a: T @@ Greater[N], b: T @@ Greater[M])(implicit sum: Sum[N,M]): T @@ Greater[sum.Out]` but was unable to apply that to `NonNegative`. Possibly because `NonNegative = Not[Less[_0]]`.

wemrysi commented on core/src/test/scala/quasar/fs/NumericGen.scala:
```diff
@@ -3,9 +3,14 @@ package quasar.fs
+  // TODO: Get rid of .get once we update versions of scalacheck
   implicit val arbPositive  = Arbitrary.apply(Gen.choose(1, Short.MaxValue).map(short => Positive(short).get))
-  implicit val arbNatural   = Arbitrary.apply(Gen.choose(0, Short.MaxValue).map(long => Natural(long).get))
+  implicit val arbNatural: Arbitrary[Long @@ NonNegative]   = Arbitrary.apply(Gen.choose[Long](0, Short.MaxValue).map(long => refineT[NonNegative](long).right.toOption.get))
```
This should probably be moved to `quasar.fp.numeric`.

jedesah replied:
Yes. Along with `Positive` (if not translated to refined). I just didn't want to spend too much time on this if the conclusion was a failed experiment. But if we think this PR is an improvement over the current state of affairs, I am glad to make those changes.

wemrysi replied:
If we're able to assign a name to the macro constructor, like I asked about above, then I think it is a good improvement. I like that the framework incorporates a sort of "liskov substitutability" for predicates. It'd just be nice not to have to expose the details of the implementation everywhere.

wemrysi commented:
One general comment, can this be rebased onto master? I'd prefer only to merge rest-api specific changes in this branch and everything else to master.

jedesah commented:
Ya :smile:. This was just like one more thing before creating that patch for master with the changes in `core`.

wemrysi commented on web/src/test/scala/quasar/api/services/CompileAndQueryServiceSpec.scala:
```diff
@@ -59,7 +63,7 @@ class CompileAndQueryServiceSpec extends Specification with FileSystemFixture wi
     val baseReq = Request(uri = baseURI, method = Method.POST)
     val req = query.map { query =>
-      val uri = baseURI.+??("offset", query.offset.map(_.shows)).+??("limit", query.limit.map(_.shows))
+      val uri = baseURI.+??("offset", query.offset.map(_.toString)).+??("limit", query.limit.map(_.shows))
```
Any reason we can't stick with `shows`?

jedesah replied:
We might have to define a `Show[A @@ NonNegative]`, but it should be possible, once again, just trying to get an idea for what it might look like as quickly as possible, but I will provide the necessary `Show` instance in order to revert back to `shows`.

jedesah commented:
@wemrysi Just updated this PR. I think I addressed most of your feedback. There are a couple places where I needed to put `Long @@ NonNegative` in order for the refined macros to work but I think it's an improvement over all. At least the PR removes more code than it adds! :smile:

Let me know what you think and then I can merge and finally create that patch for master.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,32 @@
+  type Natural = Long @@ NonNegative
+  type Positive = Long @@ RPositive
+
+  def Positive(a: Long): Option[Positive] = refineT[RPositive](a).right.toOption
```
Does this mean you weren't able to "rename" the macro-based constructors?

jedesah replied:
I am not sure what you mean.

wemrysi replied:
Something like, (from my previous comment)

`val Natural = refineMT[NonNegative].apply[Long]`

My critique here is that one of the big advantages of `refined` is the compile-time validation of literals. However, it looks like, in order to use this feature, we have to expose the inner-workings to the entire codebase (i.e. all the call sites have to either know to import some implicit macro, or you have to use the `refineMT` function explicitly), rather than being able to encapsulate this behind a constructor `Long => Natural`. The tradeoff here isn't very clear cut, we're definitely gaining something, but there is a cost as well.

jedesah replied:
I am not sure I understand how `val Natural = refineMT[NonNegative].apply[Long]` is different form `def Natural(a: Long): Option[Natural] = refineT[NonNegative](a).right.toOption`. This is for the case where it is not a literal and needs to be refined at runtime.

As for literals, I understand better what you mean. I am not convinced it's a super big deal to expose this, as it's the same import whether you are refining a number or a string or a regex, and I am not sure how reasonable it is to create a specialized macro for each refinement type me might ever want to use as opposed to just saying that if you need to lift a literal into any kind of refinement, you need an import for that.

wemrysi replied:
> I am not sure I understand how val Natural = refineMT[NonNegative].apply[Long] is different form def Natural(a: Long): Option[Natural] = refineT[NonNegative](a).right.toOption. This is for the case where it is not a literal and needs to be refined at runtime.

I wasn't advocating using it for non-literals, I was advocating for it in the case of literals, i.e. `Natural(0L)`, but it looks like that doesn't work.

wemrysi commented on core/src/main/scala/quasar/fs/ReadFile.scala:
```diff
@@ -38,7 +42,7 @@ object ReadFile {
       * the maximum amount of data read.
       */
-    def scan(file: AFile, offset: Natural, limit: Option[Positive]): Process[M, Data] = {
+    def scan(file: AFile, offset: Long @@ NonNegative, limit: Option[Positive]): Process[M, Data] = {
```
Possible to use the alias here?

jedesah replied:
For a reason I don't totally understand, assigning a `Long` literal to a `Natural` does not compile, but to a `Long @@ NonNegative` does compile even though `Natural = Long @@ NonNegative`. Putting this here avoids having to ascribe all the literals in the functions that call this one.

wemrysi replied:
:disappointed:

wemrysi commented on core/src/test/scala/quasar/fp/numeric/Gen.scala:
```diff
@@ -0,0 +1,26 @@
+
+package object scalacheck {
+  // TODO: Get rid of .get once we update versions of scalacheck
+  implicit def positive[T:Choose](implicit num: Numeric[T]): Arbitrary[T @@ RPositive] =
```
I'd prefer using either context bounds or implicit parameters for a single type, i.e. `positive[T : Choose : Numeric]`, rather than both.

jedesah replied:
I am kind of curious why that is? I feel like in this case it's the most concise way to do things here as I need access to the `Numeric` instance for `T` as opposed to `Choose` which I don't need to refer to explicitly, why create an identifier for it?

wemrysi replied:
Clarity and style mostly, I think it is clearer to have all the constraints in one spot for a given type. However, if you and the rest of the team disagree, then I'll go with whatever the consensus is.

sellout replied:
I remember hearing (via @stew?) that non/cats has discussed this particular style issue a lot. And if they’ve arrived at a conclusion, I’m inclined to adopt it. But I don’t see any style stuff on the http://non.github.io/cats

stew replied:
@sellout et. al: We had informally decided that for implicits which are NOT used in the method body, we would use context bounds. For implicits which ARE used in the method body, we would name the implicit parameter instead of using implicitly to then conjure the value inside the method body. My recollection again is that this was an informal decision that happened over gitter.

We did also discuss the name of the implicit parameter when we were naming it, and it was pretty widely agreed upon that although many of us did prefer to name this value parameter with the same name as the type parameter, using the same name in both value lan and type land was confusing to people newer to these types of things. That is to say that we would use something like:

    def foo[F[_]](implicit ev: Monad[F])

instead of:

    def foo[F[_]](implicit F: Monad[F])

as you guys already have it here.

wemrysi commented on web/src/main/scala/quasar/api/services/data.scala:
```diff
@@ -74,7 +78,7 @@ object data {
         handleOffsetLimitParams(offsetParam,limitParam){ (offset, limit) =>
           val requestedFormat = MessageFormat.fromAccept(req.headers.get(Accept))
-          download(requestedFormat, path, offset.getOrElse(Natural._0), limit)
+          download(requestedFormat, path, offset.getOrElse(0L: Long @@ NonNegative), limit)
```
Possible to use the alias?

wemrysi replied:
Oh, is this one of the cases where you have to use the implicit macro?

jedesah replied:
Ya, and that macro seems to have trouble with type synonyms.

wemrysi replied:
The examples on the `refined` github page seem to indicate that use of the `refineMT` macro function works with aliases, so we'd trade the type for the function name. Not sure which is better.

wemrysi commented on web/src/test/scala/quasar/api/zip.scala:
```diff
@@ -47,8 +55,8 @@ class ZipSpecs extends Specification with ScalaCheck with ScalazMatchers {
       def count(is: java.io.InputStream): Natural = {
-        def loop(n: Natural): Natural = if (is.read ≟ -1) n else loop(n + Natural._1)
-        loop(Natural._0)
+        def loop(n: Long @@ NonNegative): Natural = if (is.read ≟ -1) n else loop(add(n, 1L))
```
Alias?

jedesah replied:
Idem. This is the most succinct way to get the conversion from literal to refined to work.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,32 @@
+  def add[T:scala.Numeric](a:T @@ NonNegative, b: T @@ NonNegative): T @@ NonNegative =
+    refineT[NonNegative].force(implicitly[scala.Numeric[T]].plus(a,b))
+
+  implicit def equalNonNegative[T]: Equal[T @@ NonNegative] = Equal.equalA[T @@ NonNegative]
```
Rather than resorting to `equals`, can we redefine as `nonNegativeEqual[T: Equal]: Equal[T @@ NonNegative]`? (similarly for the rest and the `Show` instances).

jedesah replied:
Ya, good idea!

jedesah replied:
Actually, not really possible. Triggers a diverging implicit resolution.

wemrysi replied:
Probably because `A @@ B <: A`, :disappointed:

jedesah replied:
Ya probably, although I can't seem to wrap my mind around exactly why that would trigger a diverging implicit. And that relationship is rather useful in other contexts.

wemrysi replied:
Hm, then this isn't any better than `Object#equals`. If I have a `Equal[T]` available I'd expect `Equal[T @@ Refinement]` to use it instead of `equals`, if anything.

Something like `implicit def nonNegativeEqual[A <: AnyVal]: Equal[A @@ NonNegative]` would be better, but still not ideal in the presence of value classes. May just want to create instances for the ones we're using if we can't constrain the type.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,32 @@
+import eu.timepit.refined.refineT
+import scalaz.{Equal, Show}
+
+package object numeric {
```
I know this isn't a lot of code, but I think I'd still prefer to see the instances and the constructors, if they're going to be functions, in a file per-type. You could still have an `implicits` object here that would mixin all the individual instances for omnibus imports, but it'd be good to shy away from putting everything in the package object for organizational purposes and also to allow for a la carte imports.

jedesah replied:
Not super enthusiastic about this, but sure. Will wait for clarification on what you mean by
> if they're going to be functions

as it seems to be related to my confusion in a previous comment

wemrysi replied:
I meant something like

```scala
object Natural extends NaturalInstances {
  def apply(n: Long): Option[Natural] = ...
}

trait NaturalInstances {
  ...
}
```

and then

```scala
package object numeric {
  type Natural = Long @@ NonNegative
  type Positive = Long @@ RPositive

  object implicits extends NaturalInstances with PositiveInstances
}
```

What are your reservations?

jedesah replied:
I guess that my reservation is that these implicits are more generic than our definition of `Positive` and `Negative`. `Positive` is a `Long @@ NonNegative`, but the instance of `Show` is for any `NonNegative` `T`.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,32 @@
+  def Positive(a: Long): Option[Positive] = refineT[RPositive](a).right.toOption
+  def Natural(a: Long): Option[Natural] = refineT[NonNegative](a).right.toOption
+
+  def add[T:scala.Numeric](a:T @@ NonNegative, b: T @@ NonNegative): T @@ NonNegative =
```
`Monoid`?

jedesah replied:
I had tried, but was not able to get it to work. Can try again quick tmr.

wemrysi replied:
Any luck?

wemrysi commented on core/src/main/scala/quasar/fp/numeric/Bounded.scala:
```diff
@@ -0,0 +1,18 @@
+
+import quasar.Predef._
+
+trait Bounded[T] extends {
```
`extends`?

jedesah replied:
What?! That compiles? I had created a `LowerBound` and `UpperBound` typeclass before but ran into some compile time implicit issues (which were potentially unrelated) and decided to go for the more straightforward approach.

wemrysi commented on it/src/test/scala/quasar/fs/ReadFilesSpec.scala:
```diff
@@ -95,44 +104,41 @@ class ReadFilesSpec extends FileSystemTest[FileSystem](FileSystemTest.allFsUT) {
 
-      "scan with offset k > 0 and no limit skips first k data" >> {
-        val k = Natural._9 * Natural._2
+      "scan with offset k > 0 and no limit skips first k data" ! prop { k: Natural =>
```
Instead of limiting all of these to one case, can we just summon an arbitrary value directly, i.e. `val k = Arbitrary[Natural]`?

wemrysi replied:
Also, the meaning of this tests has changed as `k` is expected to be strictly greater than zero, `Natural` doesn't guarantee this.

jedesah replied:
Nice catch on the `k >  0`. That was an oversight on my part, but now that I think about it, I am not clear why `0` is a value we would want to avoid testing unless it's being explicitly checked in another test (which may be the case).

jedesah replied:
As for `val k = Arbitrary[Natural]`, I tend to prefer having/leaving nice pure properties and adding practical (space and time) limitations as settings at the end of the property definition. It's the same reason why I used custom generators to produce values of `Positive` that are reasonable in size instead of specifying the input type of as an `Interval` in some of these other properties.

But I guess that might be somewhat of a style consideration and we should probably strive for a uniform style, so maybe we can open up the discussion to the rest of the team and see what they think.

jedesah replied:
Also, if we summon an `Arbitrary[A]`, the generator is not guaranteed to produce a value so we would need some kind of code to try the `Gen` until it produces a `Some` which seems unpleasant to deal with. Maybe you had another approach in mind?

wemrysi replied:
`k = 0` is being tested explicitly in another example. As for arbitrary, I forgot that you don't get an actual value with `implicit[Arbitrary[Foo]]`, so converting to properties is fine.

jedesah replied:
I guess this brings back up the question of whether we could squash down the properties into one more general one. I think scalacheck is pretty good at generating values like 0 ± 1 (if in domain), `Bounded.minValue` (+1) `Bounded.maxValue` (-1), so I am not convinced that having separate properties is a better approach, but I am okay with the status quo.

wemrysi replied:
A reasonable question, but I don't think changing the semantics of the test is relevant to this PR. If we want to do this, it should be in a focused commit of its own.

jedesah replied:
Fair enough.

wemrysi commented:
So, I'm on the fence now whether we should try and use any aliases for refined types as we can't consistently apply them. For consistency's sake, I'm leaning towards just ditching them and forcing using the tagged type everywhere.

jdegoes commented:
Maybe this is not true, but based on all the comments I can't say this is a win. Looks like a slight loss and adds another sketchy dependency on macros that may or may not work in the future.

wemrysi commented:
@jdegoes I'm inclined to agree, but would welcome comments from one or more of @sellout, @mossprescott and @drostron.

jedesah commented:
I was not sure if this was a positive change at first, but now I am fairly convinced it is. These are the reasons I believe it is a positive change:

- It decouples `Natural` and `Positive` from `Long`. There are a few tests where we rely on `List.take` and `List.drop` which take `Int` in order to implement the property and we encounter failures if we do not provide custom generators for `Positive` and `Long` that generate values below `Int.MaxValue`. Now we can express the properties more correctly and avoid relying on the more fragile approach of custom generators.
- It removes code. Although currently, the PR adds more code than it removes, a large body of code can be removed from Quasar once it is merge and released upstream. I already opened a PR against refined last night and will complete the PR on my personal time. Also, many other additions are also just general improvements to the code base I would probably want to rescue even if this PR were to get shot down.
- Somewhat related to the previous point, we can use arbitrary constant literals instead of definning a bunch of code and still needing to rely on weirdness like [this](https://github.com/quasar-analytics/quasar/blob/master/it/src/test/scala/quasar/fs/ReadFilesSpec.scala#L98) in order to declare compile time constant values.
- It allows us to get rid of property preconditions by specifying the types of property inputs more precisely, ie. ```Greater[W.`100L`.T]```.

@wemrysi Also, refined has a `Refined` type that is another way to "refine" a type and would avoid the issue with type aliases not working. @fthomas actually recommended that we use `Refined` instead of `@@` if possible. If I have managed to convince you guys, then maybe we can explore using that instead but I am not too inclined to spend (even personal) time perfecting this if it's going to be shot down.

However, I would appreciate concrete reasons.

fthomas commented:
Let me comment on the differences between `@@` and `Refined`. `@@` is defined such that the refined type `T @@ P` is a subtype of the base type `T`, so `(T @@ P) <: T`. Unfortunately `@@` does not work well in combination with type aliases due to a scalac bug ([SI-8740](https://issues.scala-lang.org/browse/SI-8740)). `Refined` on the other hand is a standard value class such that there is no subtyping relation between `T` and `T Refined P` but it has no problems with type aliases. So if you want to keep using `Natural` as name, I'd recommend to define it as `type Natural = Long Refined NonNegative`.

The alias for the refine macro (so that `Natural(0L)` is the same as `refineMT[NonNegative](0L)`) that @wemrysi mentioned is an interesting idea I haven't thought of yet. I'll see if refined can provide support for that.

wemrysi commented on core/src/main/scala/quasar/fp/numeric/package.scala:
```diff
@@ -0,0 +1,29 @@
+  def Positive(a: Long): Option[Positive] = refineV[RPositive](a).right.toOption
+  def Natural(a: Long): Option[Natural] = refineV[NonNegative](a).right.toOption
+
+  implicit def widenInt[F[_,_],M](a: F[Int,M])(implicit rt: RefType[F]): F[Long,M] = rt.unsafeWrap(rt.unwrap(a).toLong)
```
I'd prefer some implicit syntax rather than an implicit conversion as the latter usually result in code that is harder to understand (i.e. when I was reading through the tests, I didn't understand how you could be passing an `Int Refined NonNegative` to something that accepted a `Natural`).

This is also a bit sketchy as it places no constraints on `M`, so we're saying _any_ predicate about an `Int` will always be true about a `Long`. For the existing numeric predicates, this looks to be true, but what about something like `eu.timepit.refined.generic.Equal`?

jedesah replied:
I think because there is an implicit conversion from `Int => Long`, that it's not unreasonable to have `Int Refined M` to `Long Refined M`. Although after writing that, I realize that it would be more appropriate to demand evidence of a conversion from `Int => Long` or from `A:Numeric => B:Numeric` at that point. Anyway, I am okay with explicit implicit syntax as well.

Good point about the `M`. I don't think there is a good way to abstract over any Numeric refinement right now, so probably need to define two case, one for `NonNegative` and one for `Positive`.

wemrysi commented on core/src/test/scala/quasar/fp/numeric/Gen.scala:
```diff
@@ -0,0 +1,74 @@
+
+import scala.Numeric
+
+import org.scalacheck.Gen.Choose
```
Minor comments, but mind organizing imports according to our style and also picking a consistent way of handling long function signatures?

jedesah replied:
Ya, you are right, imports are a mess in the PR currently. I need to familiarize myself if we have an official "style". The function signatures adopt the style in the refined codebase. I will standardize them.

wemrysi commented on it/src/test/scala/quasar/fs/ReadFilesSpec.scala:
```diff
@@ -13,10 +15,18 @@ import pathy.Path._
 import scalaz.stream._
 
-class ReadFilesSpec extends FileSystemTest[FileSystem](FileSystemTest.allFsUT) {
+import eu.timepit.refined.auto._
```
Import style

wemrysi commented on it/src/test/scala/quasar/fs/ReadFilesSpec.scala:
```diff
@@ -95,44 +104,41 @@ class ReadFilesSpec extends FileSystemTest[FileSystem](FileSystemTest.allFsUT) {
 
-      "scan with offset k > 0 and no limit skips first k data" >> {
-        val k = Natural._9 * Natural._2
+      "scan with offset k > 0 and no limit skips first k data" ! prop { k: Int Refined NonNegative =>
```
This still isn't testing what the description says, `k` should be `Positive`.

jedesah replied:
Right. I didn't change it yet because I had posted a question and was waiting for a response.

jedesah commented:
@wemrysi Thx for going through the changes again and pointing out a few possible improvements. If we decide to move forward with the PR, I will fix the issues during my next 10% time slot.

wemrysi commented:
I think going with the `Refined` type is an improvement over `@@` and makes usage more consistent. It also looks like @fthomas might have a solution for the macro aliasing, which would allow us to eventually have nicer constructors. Assuming no other issues creep up, I'm ok with this assuming nobody else on the team has any dealbreaking objections. I do think we should wait for at least a third opinion just given how contentious this has been.

jedesah commented:
:+1:

jedesah commented:
I am unaware of a way of changing the target branch without closing and re-opening.
